packages/core/src/model/heatmap.ts

```diff
@@ -0,0 +1,326 @@
+		if (domain[0] < 0 && domain[1] > 0) {
+			if (Math.abs(domain[0]) > domain[1]) {
+				domain[1] = Math.abs(domain[0]);
+			} else {
+				domain[0] = -domain[1];
+			}
+		}
+
+		return domain;
+	}
+
+	/**
+	 * @override
+	 * @param value
+	 * @returns
+	 */
+	getFillColor(value: number) {
+		return this._colorScale(value);
+	}
+
+	/**
+	 * Generate a list of all unique domains
+	 * @returns String[]
+	 */
+	getUniqueDomain(): string[] {
+		if (Tools.isEmpty(this._domains)) {
+			const displayData = this.getDisplayData();
+			const { cartesianScales } = this.services;
+
+			const domainIdentifier = cartesianScales.getDomainIdentifier();
+			const mainXAxisPosition = cartesianScales.getMainXAxisPosition();
+			const customDomain = cartesianScales.getCustomDomainValuesByposition(
+				mainXAxisPosition
+			);
+
+			// Use user defined domain if specified
+			if (!!customDomain) {
+				return customDomain;
+			}
+
+			// Get unique axis values & create a matrix
+			this._domains = Array.from(
+				new Set(
+					displayData.map((d) => {
+						return d[domainIdentifier];
+					})
+				)
+			);
+		}
+
+		return this._domains;
+	}
+
+	/**
+	 * Generates a list of all unique ranges
+	 * @returns String[]
+	 */
+	getUniqueRanges(): string[] {
+		if (Tools.isEmpty(this._ranges)) {
+			const displayData = this.getDisplayData();
+			const { cartesianScales } = this.services;
+
+			const rangeIdentifier = cartesianScales.getRangeIdentifier();
+			const mainYAxisPosition = cartesianScales.getMainYAxisPosition();
+			const customDomain = cartesianScales.getCustomDomainValuesByposition(
+				mainYAxisPosition
+			);
+
+			// Use user defined domain if specified
+			if (!!customDomain) {
+				return customDomain;
+			}
+
+			// Get unique axis values & create a matrix
+			this._ranges = Array.from(
+				new Set(
+					displayData.map((d) => {
+						return d[rangeIdentifier];
+					})
+				)
+			);
+		}
+
+		return this._ranges;
+	}
+
+	/**
+	 * Generates a matrix (If doesn't exist) and returns it
+	 * @returns Object
+	 */
+	getMatrix() {
+		if (Tools.isEmpty(this._matrix)) {
+			const uniqueDomain = this.getUniqueDomain();
+			const uniqueRange = this.getUniqueRanges();
+
+			const domainIdentifier = this.services.cartesianScales.getDomainIdentifier();
+			const rangeIdentifier = this.services.cartesianScales.getRangeIdentifier();
+
+			// Create matrix (domain by range) and initalize it's values to null
+			uniqueDomain.forEach((dom: any) => {
+				const range = {};
+				// Data will be set to null by default, to signify 'missing'
+				uniqueRange.forEach((element: any) => {
+					range[element] = {
+						value: null,
+						index: -1,
+					};
+				});
+				this._matrix[dom] = range;
+			});
+
+			// Fill in user passed data
+			this.getDisplayData().forEach((d, i) => {
+				this._matrix[d[domainIdentifier]][d[rangeIdentifier]] = {
+					value: d['value'],
+					index: i,
+				};
+			});
+		}
+
+		return this._matrix;
+	}
```

Comment:
Do we absolutely need all the processing?

The answer doesn't have to be a no, but if it is then we should simplify

Reply:
I think it is good to have this processing, it only happens on chart creation but it speeds up our axes hover. Instead of iterating through entire data length (n), we can access the row/column by simply accessing the object key (constant).

Additionally, it allows us to find any cells that are empty in the matrix.

Usage can be found here - [https://github.com/carbon-design-system/carbon-charts/blob/b57e14cdf3db0d02fd454ad356aa40a1e1996213/packages/core/src/components/graphs/heatmap.ts#L352](https://github.com/carbon-design-system/carbon-charts/blob/b57e14cdf3db0d02fd454ad356aa40a1e1996213/packages/core/src/components/graphs/heatmap.ts#L352)

Reply:
Updated efficiency from n^2 to 2n